lib/prelude/input/mod.dart

```diff
@@ -102,167 +105,128 @@ class Input {
       _debugPrint(
           "Hover:${hoveringPointer.id} (${(hoveringPointer.history.last as PointerStateHover).raw.position}) -> Down (${hoveringPointer.position})");
     } else {
       // Create a new pointer
-      final pointer = Pointer.fromTapDownEvent(_lastPointerId++, event);
+      final pointer = Pointer(state);
       _pointers.add(pointer);
-      _debugPrint("New Down:${pointer.id} (${pointer.position})");
+      _debugPrint("New PointerDown:${pointer.id} (${pointer.position})");
     }
   }
 
-  void onLongTapDown(ex.TapDownEvent event) {
-    _longTapDowns.add(event);
-
-    // Update existing pointer or throw if not found
-    final pointer = _pointers.firstWhere(
-      (pointer) =>
-          pointer.state is PointerStateDown &&
-          pointer.pointerId == event.pointerId,
-      orElse: () => throw Exception(
-          "Long tap down event received for a pointer that is not already registered as down"),
-    );
-    pointer.pushState(PointerStateLongDown(event));
-    _debugPrint("Down:${pointer.id} -> LongDown (${event.canvasPosition})");
-  }
-
-  void onTapUp(ex.TapUpEvent event) {
-    _tapUps.add(event);
-
-    // Update existing pointer or throw if not found
-    final pointer = _pointers.firstWhere(
+  void onPointerUp(PointerUpEvent event) {
+    // Update a down or longdown pointer, or
+    // finish a drag
+    final pointer = _pointers.firstWhereOrNull(
       (pointer) =>
           (pointer.state is PointerStateDown &&
-              pointer.pointerId == event.pointerId) ||
+              pointer.pointerId == event.pointer) ||
           (pointer.state is PointerStateLongDown &&
-              pointer.pointerId == event.pointerId),
-      orElse: () => throw Exception(
-          "Tap up event received for a pointer that is not already registered as down"),
+              pointer.pointerId == event.pointer) ||
+          (pointer.state is PointerStateDragStart &&
+              pointer.pointerId == event.pointer) ||
+          (pointer.state is PointerStateDragUpdate &&
+              pointer.pointerId == event.pointer),
     );
-    pointer.pushState(PointerStateUp(event));
-    _debugPrint(
-        "${pointer.history.last is PointerStateDown ? "Down" : "LongDown"}:${pointer.id} -> Up (${event.canvasPosition})");
+    if (pointer != null) {
+      if (pointer.state is PointerStateDragStart ||
+          pointer.state is PointerStateDragUpdate) {
+        final state = PointerStateDragEnd(event);
+        _dragEnds.add(state);
+        pointer.pushState(state);
+        _debugPrint(
+            "${pointer.state.debugName}:${pointer.id} -> DragEnd (${event.position.toVector2()})");
+      } else {
+        final state = PointerStateUp(event);
+        _tapUps.add(state);
+        pointer.pushState(state);
+        _debugPrint(
+            "${pointer.history.last.debugName}:${pointer.id} -> Up (${event.position.toVector2()})");
+      }
+    } else {
+      throw Exception(
```

Comment:
We might want to use an assert instead of throwing, in debug people will get notified while in "production" it will gracefully do nothing...

Reply:
Made the `assert`, but kept the `if` because it guards the type from being nullable 👍🏻